### PR TITLE
157 implement gender preferences in server

### PR DESCRIPTION
**Pull request by JoMuelle**
Currently default preference for gender is all genders (easier for testing out some stuff in the front end)

### CODE REVIEW

File: src/main/java/ch/uzh/ifi/hase/soprafs22/controller/UserController.java
Comment (daveking7) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/controller/UserController.java:196]: Why do you use GeneralAccess here?

File: src/main/java/ch/uzh/ifi/hase/soprafs22/controller/UserController.java
Comment (daveking7) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/controller/UserController.java:195]: since when do we use integer here? ;)

File: src/main/java/ch/uzh/ifi/hase/soprafs22/entity/User.java
Comment (daveking7) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/entity/User.java:77]: Use a Set here, also, you would use OneToMany for List.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UserGetPreferencesDTO.java
Comment (daveking7) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UserGetPreferencesDTO.java:37]: Does this work? Have you tested it in Postman? (I remember you having troubles with Gender in User)

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/mapper/DTOMapper.java
Comment (daveking7) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/rest/mapper/DTOMapper.java:54]: I would merge that ( = make one (bigger) UserPutDTO), and make a general updateUser method, which updates all fields that are given in the DTO. Also, does the frontend give the userId in the body? I think not (only as a path variable). So remove the "id" mapping (also in the DTO). Furthermore, since we don't allow for changing birthday, you can remove that as well.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java
Comment (daveking7) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java:56]: looks weird

File: src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java
Comment (daveking7) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java:107]: First, if I'm born in September 2000 then you would falsely assume that I'm 22, even though I'm still 21 and give me the default minAge of 18 instead of 19. Secondly, you write alot of code twice. 
I propose changing it to the following:
```
  private long getAgeInMilliSeconds(Date birthday) {
    return new Date().getTime() - birthday.getTime();
  }
  
  private int getAge(Date birthday) {
    Calendar c = Calendar.getInstance();
    c.setTime(new Date());
    int currentYear = c.get(Calendar.YEAR);
    c.setTime(birthday);
    int birthYear = c.get(Calendar.YEAR);
    return currentYear - birthYear;
  }

  private int findMinAgeDefault(Date userBirthday){
    long age = getAgeInMilliSeconds(userBirthday);
    return age - 3 * Time.ONE_YEAR < 18 * Time.ONE_YEAR ? 18 : getAge(userBirthday) - 3;
  }
  
  private int findMaxAgeDefault(Date userBirthday) {
    return getAge(userBirthday) + 3;
  }
```
You can add  `  public static final long ONE_YEAR = 31536000000L;
` in the Time File for this.
But props for not using the deprecated date.getYear()!

File: src/main/java/ch/uzh/ifi/hase/soprafs22/controller/UserController.java
Comment (daveking7) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/controller/UserController.java:210]: (See comments below about merging userPUTDTO's)

File: src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java
Comment (daveking7) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java:240]: I would propose to make one big updateUser method, and only update the values that change. I think for readability you can do something like:
`<condition> ? userToBeUpdated.updateField(user.getField()) : null`
, but better check if that really works.

**Review by daveking7 — CHANGES_REQUESTED** [review]
See comments, also, why do Github Actions fail? Can you check on that?

File: src/main/java/ch/uzh/ifi/hase/soprafs22/controller/UserController.java
Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/controller/UserController.java:196]: Fixed

File: src/main/java/ch/uzh/ifi/hase/soprafs22/controller/UserController.java
Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/controller/UserController.java:195]: Fixed

File: src/main/java/ch/uzh/ifi/hase/soprafs22/entity/User.java
Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/entity/User.java:77]: is set not, but keeping it as element collection

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/mapper/DTOMapper.java
Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/rest/mapper/DTOMapper.java]: Done

File: src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java
Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java:244]: I think i would keep the two different methods, as you update the user himself once and once his preferences. Its also easier to read this way and easier to modify

File: src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java
Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java]: implemented you methods. Thanks

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UserGetPreferencesDTO.java
Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UserGetPreferencesDTO.java]: needs adjusting. Will do it on the weekend

Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UserGetPreferencesDTO.java]: Actually seems to be working but will check again on sunday if i have time. Dont merge yet

Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UserGetPreferencesDTO.java]: Works without a problem now. Can be passed as array of strings from Client.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java
Comment (daveking7) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java:234]: Also, what happens if the client does not pass a genderPreference, i.e. does not update it? It would get assigned null? I would add some second lifeline here, so that the server also checks if the values passed are legit (I mean, if we can trick the system via Postman, everyone can). (check with something like userTUP.getMaxAge().equals(null) or smth.)

File: src/test/java/ch/uzh/ifi/hase/soprafs22/service/UserServiceIntegrationTest.java
Comment (daveking7) [inline src/test/java/ch/uzh/ifi/hase/soprafs22/service/UserServiceIntegrationTest.java:64]: Why do you set the birthday as today (would make more sense to have him be older than 18)

File: src/test/java/ch/uzh/ifi/hase/soprafs22/service/UserServiceIntegrationTest.java
Comment (daveking7) [inline src/test/java/ch/uzh/ifi/hase/soprafs22/service/UserServiceIntegrationTest.java:87]: Why do you set the birthday as today (would make more sense to have him be older than 18)

File: src/test/java/ch/uzh/ifi/hase/soprafs22/service/UserServiceTest.java
Comment (daveking7) [inline src/test/java/ch/uzh/ifi/hase/soprafs22/service/UserServiceTest.java:57]: Why do you set the birthday as today (would make more sense to have him be older than 18)

File: src/test/java/ch/uzh/ifi/hase/soprafs22/service/UserServiceIntegrationTest.java
Comment (daveking7) [inline src/test/java/ch/uzh/ifi/hase/soprafs22/service/UserServiceIntegrationTest.java:101]: Why do you set the birthday as today (would make more sense to have him be older than 18)

**Review by daveking7 — CHANGES_REQUESTED** [review]
You're getting there, but still not 100% happy with it.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UserGetPreferencesDTO.java
Comment (daveking7) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UserGetPreferencesDTO.java]: Nice!

File: src/main/java/ch/uzh/ifi/hase/soprafs22/controller/UserController.java
Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/controller/UserController.java:199]: Yes, it makes more sense to merge the two. For the edit preferences (Where you only need the preferences) i dont think it should be a problem if we pass "too much".

Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/controller/UserController.java]: The mapping now is just with the user getDTO (i left the API call for getpreferences, with a TODO to delete so it keeps working until the client is also changed (i am currently working on a Task where all the stuff is also adjustes in the client to unly us getUser and not getUserPreferences))

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UserGetPreferencesDTO.java
Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UserGetPreferencesDTO.java:39]: not anymore, deleted

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UserPutDTO.java
Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UserPutDTO.java:10]: yes, didn't see it. Guess it was there before

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/mapper/DTOMapper.java
Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/rest/mapper/DTOMapper.java:42]: done

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/mapper/DTOMapper.java
Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/rest/mapper/DTOMapper.java:31]: done

File: src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java
Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java:234]: I don't save and flush (actually we also don't on "update user". Should we?

Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java:234]: For min and max age you can't pass null, as they are ints

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UserPutDTO.java
Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UserPutDTO.java:10]: I put the ID back. For some reason it is used in a lot of the tests (i created a Task in the Bugfix issue, to take care of it)

File: src/test/java/ch/uzh/ifi/hase/soprafs22/service/UserServiceIntegrationTest.java
Comment (JoMuelle) [reply src/test/java/ch/uzh/ifi/hase/soprafs22/service/UserServiceIntegrationTest.java:101]: In all of them it doesn't really make a difference, as it is not what we are testing here

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UserPutDTO.java
Comment (daveking7) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UserPutDTO.java:10]: I see, we shall remove it in the tests as well.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/controller/UserController.java
Comment (daveking7) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/controller/UserController.java:214]: might be a bit confusing to call the (unreal) user here userPreferences and in updatePreferences just user and the true entity userToUpdatePreferences.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java
Comment (daveking7) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java:234]: if user.getGenderPreferences() is null, the first condition evaluates to false and the second is checked, which would result in an error, since null does not have a method isEmpty(). I think what you want is 
`!Objects.isNull(user.getGenderPreferences()) && !user.getGenderPreferences().isEmpty()`
or equivalently
`! (Objects.isNull(user.getGenderPreferences()) || user.getGenderPreferences().isEmpty())`

**Review by daveking7 — CHANGES_REQUESTED** [review]

**Comment by daveking7** [comment]
Before you request another pull, can you merge dev into this branch?

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UserPutDTO.java
Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/UserPutDTO.java:10]: Yes, will be done in a different task, so backend and frontend should both be up to date soon. Will merge dev into this tonight the PR hopefully is good

File: src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java
Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java]: Updated using :!Objects.isNull(user.getGenderPreferences()) && !user.getGenderPreferences().isEmpty()

File: src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java
Comment (JoMuelle) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/service/UserService.java:72]: ahh yes thanks

**Pull request merged**
